Recognize timezoned labels when accessing dataframes.

1kastner opened the pull request:
- [X] closes #16785 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

1kastner committed:
Recognize timezoned labels when accessing dataframes.

1kastner commented:
Ouch, looks like some for me...

jschendel commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +123,30 @@ def test_consistency_with_tz_aware_scalar(self):
         assert result == expected
 
+    def test_access_datetimeindex_with_timezoned_label(self):
+
```
Can you add the github issue number here as a comment?  (see the test after this one for an example)

1kastner replied:
Of course

1kastner replied:
Added now.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1273,52 +1273,57 @@ def _parsed_string_to_bounds(self, reso, parsed):
         lower, upper: pd.Timestamp
 
         """
+        if parsed.tzinfo is None:
```
so you need to do something different here. leave the dates that are generated in the current tz (e.g. ``self.tz``). Then you need to convert it to the parsed_tz, BUT then you need to localize back into the original timezone. There are a number of cases, here's an example.

```
# index in US/Eastern, parsed in UTC
In [22]: pd.Timestamp('20130101',tz='US/Eastern')
Out[22]: Timestamp('2013-01-01 00:00:00-0500', tz='US/Eastern')
00:00+0000', tz='UTC')

In [24]: pd.Timestamp('20130101',tz='US/Eastern').tz_convert('UTC').tz_localize(None).tz_localize('US/Eastern')
Out[24]: Timestamp('2013-01-01 05:00:00-0500', tz='US/Eastern')

# index is naive, parsed is UTC, ineffect no change here
In [25]: pd.Timestamp('20130101')
Out[25]: Timestamp('2013-01-01 00:00:00')

In [26]: pd.Timestamp('20130101').tz_localize('UTC').tz_localize(None)
Out[26]: Timestamp('2013-01-01 00:00:00')
```

As I am writing this, it looks overly complicated. I might choose instead to raise if the timezones don't match (they can be same tz or both None).

1kastner replied:
As I elaborated in #16785 the timezones do not need to match. There are quite usual cases with daylight savings time that require some flexibility here. So my idea is to change `target_tz = parsed.tzinfo` to some kind of conversion which you mentioned. Next week I might give it a shot. Dealing with only the timezone of the DatetimeIndex seems reasonable.

1kastner replied:
I am really tired but tried to find a solution. If it works, great, if not, at least it shows the idea. I will come back to it as soon as possible. Now I am having some issues with installing the development environment under Windows 10 and all solutions found look quite time consuming. Sorry when I spam you with not working code.

jreback commented:
pls rebase when you can

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

pep8speaks commented:
Hello @1kastner! Thanks for updating the PR.



 - In the file [`pandas/tests/indexing/test_datetime.py`](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py), following are the PEP8 issues :

> [Line 128:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L128): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 131:37](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L131): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 137:51](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L137): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 142:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 144:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L144): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 145:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L145): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 146:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 147:80](https://github.com/pandas-dev/pandas/blob/229783358aea09a6d494955549c3524ae1c58401/pandas/tests/indexing/test_datetime.py#L147): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)

Commit:
Make `test_access_datetimeindex_with_timezoned_label` PEP08 compliant.

Commit:
add translate function for converting time zones.

1kastner commented:
I did a rebase and some more minor adjustments.

jbrockmendel committed:
Move NaT to self-contained module (#18014)

jbrockmendel committed:
Separate out arithmetic tests for datetimelike indexes (#18049)

datapythonista committed:
Adding skip to test failing because of lxml import (#17747) (#17748)

jbrockmendel committed:
a zillion flakes (#18046)

jreback committed:
TST: separate out grouping-type tests (#18057)

GuessWhoSamFoo committed:
BUG: DataFrame.groupby() interprets tuple as list of keys

closes #17979

Author: sfoo <sfoohei@gmail.com>
Author: Jeff Reback <jeff@reback.net>

Closes #17996 from GuessWhoSamFoo/groupby_tuples and squashes the following commits:

afb00317a [Jeff Reback] TST: separate out grouping-type tests
c52b2a88d [sfoo] Moved notes to 0.22; created is_axis_multiindex var - pending internal use
fb52c1c97 [sfoo] Added whatsnew; checked match_axis_length
99ebc4ec2 [sfoo] Cast groupby tuple as list when multiindex

jreback committed:
CLN: some lint issues

LiamIm committed:
read_html(): rewinding [wip] (#18017)

jreback committed:
CI: temp disable scipy on windows 3.6 build (#18078)

mattayes committed:
DOC: Remove duplicate 'in' from contributing.rst (#18040) (#18076)

topper-123 committed:
improve test output for Categoricals (#18069)

gfyoung committed:
MAINT: Remove np.array_equal calls in tests (#18047)

jbrockmendel committed:
Move scalar arithmetic tests to tests.scalars (#18075)

TomAugspurger committed:
Update Contributing Environment section (#18052)

jbrockmendel committed:
Index tests in the wrong places (#18074)

jbrockmendel committed:
Move comparison utilities to np_datetime; (#18080)

jbrockmendel committed:
Separate _TSObject into conversion (#18060)

jbrockmendel committed:
Port Timedelta implementation to tslibs.timedeltas (#17937)

jreback committed:
COMPAT: compare platform return on 32-bit (#18090)

xref #18047

manrajgrover committed:
Fix 18068: Updates merge_asof error, now outputs datatypes (#18082)

Licht-T committed:
TST: Add regression test for empty DataFrame groupby (#18097)

Licht-T committed:
BUG: Fix the error when reading the compressed UTF-16 file (#18091)

Licht-T committed:
BUG: Implement PeriodEngine to fix PeriodIndex truncate bug (#17755)

jbrockmendel committed:
standardize indentation, arrange in allphabetical order (#18104)

gfyoung committed:
BLD: Make sure to copy ZIP files for parser tests (#18108)

Follow-up to gh-18091.

jreback committed:
Revert "CI: temp disable scipy on windows 3.6 build (#18078)" (#18105)

* Revert "CI: temp disable scipy on windows 3.6 build (#18078)"

This reverts commit cd6dc87466e119aabb76d8439df8289d082ea948.

* use numpy=1.13

jbrockmendel committed:
Masking and overflow checks for datetimeindex and timedeltaindex ops (#18020)

closes #17991

gfyoung committed:
BUG: Override mi-columns in to_csv if requested (#18110)

Previously, MultiIndex columns weren't being
overwritten when header was passed in for to_csv.

Closes gh-5539

1kastner commented:
Depends on #18095

1kastner commented:
@jreback To be frank I am rather new to this kind if github set-up and with the rebase now it looks like my pull request is a bit blown up. Is that usual or were some of my git commands not in line with what this setup is optimized for?

1kastner committed:
fix failing tests.

1kastner committed:
Merge branch 'error-on-non-naive-datetime-strings' of https://github.com/1kastner/pandas into error-on-non-naive-datetime-strings

# Conflicts:
#	pandas/core/indexes/datetimes.py
#	pandas/tests/indexing/test_datetime.py

jreback commented:
@1kastner you can

``git rebase -i upstream/master`` to rebase

alternatively you can
``git merge upstream/master`` will pretty much do the same thing

jreback commented:
> @jreback How should the whatsnew entry look like? What I added is that if the datetime index has timezone set and the label used for indexing has a timezone provided then that timezone is considered. The question at hand now is: Is that sufficient?

> My idea: if the label contains a timezone but the datetime index does not support that, there should be a user warning. An error seems harsh because maybe the user knows what s/he is doing. Do you think that is a good idea? Or should we just silently ignore the timezone as it was done before?

If the user provides a timezone, then that should be used for indexing. If the target *has* a timezone (including UTC), then this is an unambiguous operation.

If the target timezone is naive, we *could* treat it as UTC for indexing purposes. I think this would be ok (though a warning in this case would be fine).

1kastner commented:
@jreback My idea was that if no timezone is provided, the timezone of the datetime index is chosen. That seems more intuitive to me.

1kastner committed:
Rewrite naive/timezone matrix condition, Improve test cases

1kastner committed:
adjust as it was before (un-done changes)

1kastner committed:
Add tz keyword.

1kastner commented:
@jreback for me it looks like ready to go now.

jreback commented:
will have a look

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1265,14 +1265,32 @@ def _parsed_string_to_bounds(self, reso, parsed):
         reso : Resolution
             Resolution provided by parsed string.
-        parsed : datetime
+        parsed : datetime | object
```
-> or

1kastner replied:
sure, sorry for that typo

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1265,14 +1265,32 @@ def _parsed_string_to_bounds(self, reso, parsed):
         lower, upper: pd.Timestamp
 
         """
+        if not hasattr(parsed, "tzinfo"):
```
wrap parsed in ``Timestamp``, you are using datetime routines which may not always be applicable depending on tz's are used (e.g. if dateutil or pytz).

1kastner replied:
can you please elaborate on that?

1kastner replied:
I assume datetime objects here, so I expect the timezones to be datetime compatible. How could I wrap them in a way that this always works? Could you sketch some test cases I need to satisfy? Thanks a lot!

jreback replied:
``parsed = Timestamp(parsed)``

then you only care about ``.tz`` and can do ``.tz_convert()`` and such, much more idiomatic. we rareley deal directly with ``datetime`` objects

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -211,7 +213,11 @@ def test_stringified_slice_with_tz(self):
         idx = DatetimeIndex(start=start, freq="1d", periods=10)
         df = DataFrame(lrange(10), index=idx)
-        df["2013-01-14 23:44:34.437768-05:00":]  # no exception here
+        with warnings.catch_warnings(record=True) as w:
```
use ``tm.assert_raises_warning``

1kastner replied:
I pulled the latest changes from master but i could not find a `tm.assert_raises_warning`, there is only a `tm.assert_raises_regex` which is meant for errors, not warnings and a  `assert_produces_warning` which is most suitable for this situation. I will take that one.

jreback commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +125,74 @@ def test_consistency_with_tz_aware_scalar(self):
         result = df[0].at[0]
         assert result == expected
 
+    def test_access_timezoned_datetimeindex_with_timezoned_label(self):
```
these should be in test_partial.py

1kastner replied:
is moved

jreback commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +125,74 @@ def test_consistency_with_tz_aware_scalar(self):
+        df = pd.DataFrame(index=first_january, data=np.arange(len(
+                          first_january)))
+
+        former_naive_endpoint_df = df[
```
why are you naming things like this?

1kastner replied:
do you have a better name?

1kastner replied:
I called it naive because before my implementation the datetime objects were treated like naive datetime objects even when they provided a timezone which was just dropped.

jreback commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +125,74 @@ def test_consistency_with_tz_aware_scalar(self):
+            "2016-01-01T00:00-02:00":"2016-01-01T02:03"
+        ]
+
+        former_non_naive_endpoint_df = df[
```
construct the expected return dataframe and simply compare using ``assert_frame_equal``

1kastner replied:
thanks for that hint!

jreback commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +125,74 @@ def test_consistency_with_tz_aware_scalar(self):
+            "2016-01-01T00:00-02:00"):pd.Timestamp("2016-01-01T02:03")]
+
+        assert (len(former_naive_endpoint_df.index) == 4)
+
```
same

jreback commented on pandas/tests/indexing/test_datetime.py:
```diff
@@ -123,6 +125,74 @@ def test_consistency_with_tz_aware_scalar(self):
+        df = pd.DataFrame(index=first_january, data=np.arange(len(
+                          first_january)))
+
+        with warnings.catch_warnings(record=True) as w:
```
assert_raises_warning

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

Commit:
Apply suggestions of review

Commit:
refactor: replace _utc() with utc

1kastner committed:
fix flake8 issues

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

1kastner commented:
except the timestamp wrapping the other topics have been covered in my last commits.

Commit:
replace datetime.datetime with pd.Timestamp

1kastner commented:
@jreback do you have any further feedback?

jreback commented:
lgtm. can you add a whatsnew note, maybe in Other Enhancements. Also in the partial string indexing doc section, can you add a note (or maybe a warning) about this, and show an example?

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

1kastner committed:
Add whatsnew and documentation.

1kastner committed:
Fix variable name in documentation

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,40 @@ We are stopping on the included end-point as it is part of the index
+If a timezone is provided by the label, the datetime index is assumed to be UTC and a ``UserWarning`` is emitted.
+
+.. ipython:: python
+
```
this will show the warning in the docs, so use ``:okwarning:``

1kastner replied:
done

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,40 @@ We are stopping on the included end-point as it is part of the index
+   four_minute_slice
+
+
+``DatetimeIndex`` partial string indexing can be used with naive labels as well.
```
reverse these sections (naive first)

1kastner replied:
done

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,40 @@ We are stopping on the included end-point as it is part of the index
+.. ipython:: python
+
+   first_january_cet = pd.date_range('2016-01-01T00:00', '2016-01-01T23:59',
+                                     freq='T', tz="CET")
```
show the df here (and in other section as well)

1kastner replied:
I thought it will be too long, but sure

jreback replied:
oh its minutely data, hmm, though should be constrained by our max_rows anyhow.

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,40 @@ We are stopping on the included end-point as it is part of the index
    dft2.loc[idx[:, '2013-01-05'], :]
 
+
+``DatetimeIndex`` partial string indexing also works with timezones.
```
add a versionadded tag

1kastner replied:
sure

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,40 @@ We are stopping on the included end-point as it is part of the index
    dft2 = dft2.swaplevel(0, 1).sort_index()
    dft2.loc[idx[:, '2013-01-05'], :]
 
+
```
note that this behavior works with Timestamps or strings

1kastner replied:
actually also `datetime.datetime` as `pd.Timestamp` can also digest that. Should it be a note like I did now?

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -22,7 +22,7 @@ Other Enhancements
 
 - :meth:`Timestamp.timestamp` is now available in Python 2.7. (:issue:`17329`)
--
+- :class:`DatetimeIndex` is now timezone aware when using strings as labels  (:issue:`16785`)
```
say for indexing

1kastner replied:
done, but out of interest: how do you believe this can be understood in a different way?

jreback commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,46 @@ def test_partial_set_empty_frame_empty_consistencies(self):
         df.loc[0, 'x'] = 1
         expected = DataFrame(dict(x=[1], y=[np.nan]))
         tm.assert_frame_equal(df, expected, check_dtype=False)
+
```
parametrize both of these tests with strings & with Timestamp for the indexers

1kastner replied:
either I do not understand you correctly or re-read the tests. I already both use string labels and `pd.Timestamp` labels there.

1kastner committed:
Apply review suggestions.

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -22,7 +22,7 @@ Other Enhancements
 
 - :meth:`Timestamp.timestamp` is now available in Python 2.7. (:issue:`17329`)
--
+- :class:`DatetimeIndex` is now timezone aware when using strings as labels for indexing (:issue:`16785`)
```
this is a bug fix

1kastner replied:
now it is moved

jreback commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,46 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+        df = pd.DataFrame(index=first_january, data=np.arange(len(
+                          first_january)))
+
+        former_naive_endpoint_df = df[
```
can you use result= and expected= here (and then a comment if needed)

1kastner replied:
that's simple and done

jreback commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,46 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+            former_naive_endpoint_df = df["2016-01-01T00:00-02:00":
+                                          "2016-01-01T02:03"]
+
+        former_non_naive_endpoint_df = df[pd.Timestamp(
```
same

1kastner replied:
that's simple and done

jreback commented:
ping when you update

1kastner committed:
Move change to bug and rename into result and expected

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

1kastner commented:
@jreback ping on update

1kastner commented:
The failing test I accidentially got from the master branch.

jbrockmendel commented:
@jreback pls hold off on this pending resolution of #18435.

jreback commented:
@jbrockmendel your discussion w.r.t. to this issue is misplaced. We are talking about how to interpret partial strings when you already know the underlying details about the tz's in question. This is just user convenience and is completely orthogonal to other issues you have raised.

jbrockmendel commented:
> This is just user convenience and is completely orthogonal to other issues you have raised.

I understand the convenience issue.  Consider the two types of slicing/indexing `series[lower:upper]` and `series[(series.index >= lower) & (series.index <= lower)]`.  This equivalence breaks the orthogonality.  If you want to break the equivalence that's fine, I just want that design decision to be made explicit.

1kastner commented:
@jreback Is this pull request ready to go?

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -236,7 +236,8 @@ def test_stringified_slice_with_tz(self):
         df = DataFrame(lrange(10), index=idx)
-        df["2013-01-14 23:44:34.437768-05:00":]  # no exception here
+        with tm.assert_produces_warning(UserWarning):
+            df["2013-01-14 23:44:34.437768-05:00":]  # no exception here
```
I would remove the ``# no exception here`` and add a comment about the warning that is produced

1kastner replied:
I am not sure whether that is a good idea. That comment is not mine and it is not related to my code. The proposed refactoring is beyond the scope of this pull request.

1kastner commented:
I am not sure whether that is a good idea. That comment is not mine and it is not related to my code. The proposed refactoring is beyond the scope of this pull request.

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

jreback commented:
> I am not sure whether that is a good idea. That comment is not mine and it is not related to my code. The proposed refactoring is beyond the scope of this pull request.

@1kastner not sure what you are referring. I had only 1 more comment (small). @jbrockmendel concerns will be addressed elsewhere.

jbrockmendel commented:
@jreback I think the comment @1kastner was referring to was the `#no exception` comment you asked to have removed.  Ambiguity in code-comment vs GH comment.

> I would remove the `# no exception` here

1kastner commented:
@jreback This test checks whether no exception is thrown. Actually it is better to completely remove this test because whatever it tests is covered by my new tests as well

jreback commented:
> @jreback This test checks whether no exception is thrown. Actually it is better to completely remove this test because whatever it tests is covered by my new tests as well

that would be fine

1kastner committed:
Add CET timezoned datetime index as another test case

1kastner committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into error-on-non-naive-datetime-strings

Commit:
Adjust for flake8

jreback commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,50 @@ We are stopping on the included end-point as it is part of the index
    dft2 = dft2.swaplevel(0, 1).sort_index()
    dft2.loc[idx[:, '2013-01-05'], :]
 
+.. versionadded:: 0.21.1
```
add a sub-section label here (with a ref), call it something like ``slicing with timezones``.

jreback commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,66 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+        # GH 6785
+        # timezone was ignored when string was provided as a label
+
+        first_january = pd.date_range('2016-01-01T00:00', '2016-01-01T23:59',
```
I would really like to parametrize these to avoid the code repetition.  so i think you can do it with 2 test functions, one which slices and compares with an expected, and the 2nd function which checks for the warnings (you can actually do it with one if you add some more paramaters)

something like

```
@pytest.mark.parametrize("tz, start, end",......)
```

1kastner replied:
I think this contradicts the idea of having the `df.iloc` check suggested by @jorisvandenbossche because that is rather specific. I would rather delete the non-naive UTC test because the CET test shows much more.

jorisvandenbossche replied:
You should be able to write the different strings so that they give the same expected frame I think

jorisvandenbossche commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1311,14 +1311,28 @@ def _parsed_string_to_bounds(self, reso, parsed):
+                parsed = parsed.tz_convert(utc)
+        else:
+            if parsed.tz is None:  # treat like in same timezone
+                parsed = parsed.tz_localize(self.tz)
```
This case already worked before AFAIK, do you know why this is needed? (although the code seems logical)

1kastner replied:
The code is not necessarily needed when it is done somewhere else

jorisvandenbossche commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,66 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+
+        result = df[
+            "2016-01-01T00:00-02:00":"2016-01-01T02:03"
+        ]
```
Can you put this all on a single line?

jorisvandenbossche commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,66 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+            pd.Timestamp("2016-01-01T02:03")
+        ]
+
+        tm.assert_frame_equal(result, expected)
```
Can you assert both results (with strings or with Timestamps) with an independelty constructed one? (eg `df.iloc[...]`)

1kastner replied:
I am not sure what you mean with `df.iloc[...]` but generally speaking yes

jorisvandenbossche replied:
I mean to created 'expected' with something like df.iloc[120:124] (but then with the cirrect numbers)

jorisvandenbossche commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,66 @@ def test_partial_set_empty_frame_empty_consistencies(self):
         expected = DataFrame(dict(x=[1], y=[np.nan]))
         tm.assert_frame_equal(df, expected, check_dtype=False)
+
+    def test_access_timezoned_datetimeindex_with_timezoned_label_utc(self):
```
I think this is the incorrect file for the tests, as the partial in `test_partial.py` is not referring partial datetime string indexing, but partial setting or something. 
There is a `indexes/datetimes/test_partial_slicing.py` that seems a better fit.

1kastner replied:
Possible, this was a suggestion of @jreback.

jorisvandenbossche replied:
Yes, but I think it was a typo of @jreback and he forgot the _slicing part. You can look in the current file and see that there is no test related to datetime index slicing, so please make the change

jorisvandenbossche commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -637,3 +637,66 @@ def test_partial_set_empty_frame_empty_consistencies(self):
+
+    def test_access_timezoned_datetimeindex_with_timezoned_label_utc(self):
+
+        # GH 6785
```
I think this is an incorrect issue number

1kastner replied:
oh let me have a look how that could happen

jorisvandenbossche reviewed (commented):
I would leave this for 0.22.0 instead of 0.21.1 (since it raises some discussion, I think it is good to have it in master for some time longer)

I think I am in general fine with the changes in this PR (as it is aligning `df[lb:up]` with `df[df.index >= lb & df.index <= up]` which already worked with timezones), but I agree with @jbrockmendel that the discussion in https://github.com/pandas-dev/pandas/issues/18435 *is* relevant. I think that we should try to have comparisons, scalar element accessing and partial slicing all work consistent (which is however the case in this PR, so I fine with going forward with this PR is it is for 0.22.0, and we can continue to discuss on the other issue about the more broader aspect).  
Something else related to this:

- If we add a warning when doing partial string indexing with aware string on naive index, we should do the same for accessing a scalar with a string (eg `df.loc["2016-01-01T00:00-02:00"]` with your example in the docs)

jorisvandenbossche commented:
@1kastner Can you also test with an implicit partial string slice? (I mean a single string that represents a slice, without actually slicing, so eg `df["2016-01-01T00-02:00"]`)

jbrockmendel commented:
The slicing code goes through a different branch for monotonic vs non-monotic DatetimeIndexes.  Is there a test for this that goes through that path?

jbrockmendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -1364,7 +1378,7 @@ def _parsed_string_to_bounds(self, reso, parsed):
                           parsed.hour, parsed.minute, parsed.second,
                           parsed.microsecond)
-            return (Timestamp(st, tz=self.tz), Timestamp(st, tz=self.tz))
+            return Timestamp(st, tz=self.tz), Timestamp(st, tz=self.tz)
```
Since you're already making edits here there's a small bug-like that might be worth fixing.  The day, hour, minute, and second cases don't have `tz=self.tz` passed to the upper half of the returned tuple.

1kastner replied:
Thanks for pointing that out! By now the timezone is like ignored (or at least not considered well enough) in the later stages but maybe one day that will change.

1kastner commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,50 @@ We are stopping on the included end-point as it is part of the index
+
+.. note::
+
+   This both works with ``pd.Timestamp`` and strings
```
Please talk to @jreback who suggested to mention it. Actually it also works for `datetime.datetime`.

1kastner commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,50 @@ We are stopping on the included end-point as it is part of the index
+.. ipython:: python
+   :okwarning:
+
+   first_january_implicit_utc = pd.date_range('2016-01-01T00:00', '2016-01-01T23:59',
```
It can be shortened but I would keep it a bit longer than the first 10 because of the comparison in the end.

jorisvandenbossche replied:
Which comparison?

1kastner replied:
I thought (without carefully checking) that maybe in the end I will just compare two empty dataframes which will accidentially happen to be equal. To avoid such wrong positive test I thought having a bit longer df can be helpful.

jorisvandenbossche replied:
This are the docs, not tests. And you perfectly control what you do in the example, so you can just make it a bit longer than needed for the slicing to see the effect.

1kastner commented on doc/source/timeseries.rst:
```diff
@@ -557,6 +557,50 @@ We are stopping on the included end-point as it is part of the index
+
+   df
+
+   four_minute_slice = df["2016-01-01T00:00-02:00":"2016-01-01T02:03"]
```
Yes you are right. What is the consequence in your eyes? I just want the timezones to work, that is my only desire.

jorisvandenbossche replied:
There is no consequence for the behaviour, so this PR will your usecase, But for the example in the docs, we should make a clear one. So either I would make this actual partial slicing, or move this section to somewhere else

1kastner replied:
Then better move it because the timezones can not always be parsed, e.g. for months still UTC will be assumed as it goes through another path.

jorisvandenbossche replied:
No, you can just edit the example a little bit. For example keep the minute resolution, and use strings with only hours (instead of the minutes now, and that still provides ability to specify time zone), or change the resolution of the df to seconds, and keep the strings as they are. Note you can do eg each 30s to avoid that selecting some minutes results in many rows.

1kastner commented:
As I need to focus on my work for now, I will not continue for a while. I set that maintainers are allowed to edit the pull request in case of urgent need for action.

jreback commented:
can you rebase. let's see where we are on this.

1kastner commented:
I just checked and it means a lot of merging. I do not have the resources to deal with this subject even though the problem itself still bugs me.
